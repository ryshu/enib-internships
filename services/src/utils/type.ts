/** @property {MentorRole[]} mentorRoles array of each mentors roles availables */
export const mentorRoles: MentorRole[] = ['default', 'admin'];

/**
 * @summary Method used to check if given string is a mentor role
 * @param {string} str string to check
 * @returns {boolean} check
 */
export function isMentorRole(str: string): str is MentorRole {
    return mentorRoles.includes(str as MentorRole);
}
