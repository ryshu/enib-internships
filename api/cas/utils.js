"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** @property {string} regex pattern to identify students */
const regex = /[a-z][0-9][a-z]{1,6}@enib\.fr/;
/**
 * @summary Method used to check if given casUser match student pattern
 * @param {string} casUser cas username
 * @returns {boolean} check result
 */
function isStudent(casUser) {
    if (regex.test(casUser)) {
        return true;
    }
    return false;
}
exports.isStudent = isStudent;
/**
 * @summary Method used to get email of given cas user
 * @param {string} casUser cas username
 * @returns {string} email
 */
function getEmail(casUser) {
    return `${casUser}@enib.fr`;
}
exports.getEmail = getEmail;
/** @summary Macro to set first letter upper case */
function ucFirst(str) {
    return str ? `${str[0].toUpperCase()}${str.slice(1)}` : '';
}
/**
 * @summary Method used to guess a mentor name
 * @param {string} casUser cas username
 * @returns {{ firstName: string; lastName: string }} First and last name guessed
 */
function guessMentorFullName(casUser) {
    const tmp = casUser.split('.');
    if (tmp.length !== 2) {
        throw new Error('Mentor name should use following pattern `firstName.lastName`');
    }
    return {
        firstName: ucFirst(tmp[0]),
        lastName: tmp[1].toUpperCase(),
    };
}
exports.guessMentorFullName = guessMentorFullName;
/**
 * @summary Method used to guess a student name
 * @param {string} casUser cas username
 * @returns {{ firstName: string; lastName: string }} First and last name guessed
 */
function guessStudentFullName(casUser) {
    if (!regex.test(casUser)) {
        throw new Error('Student name should use following pattern `/[a-z][0-9][a-z]{1, 6}/`');
    }
    return {
        firstName: casUser[0].toUpperCase(),
        lastName: casUser.slice(2).toUpperCase(),
    };
}
exports.guessStudentFullName = guessStudentFullName;
//# sourceMappingURL=utils.js.map