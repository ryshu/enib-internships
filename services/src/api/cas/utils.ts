/** @property {string} regex pattern to identify students */
const regex = /[a-z][0-9][a-z]{1,6}@enib\.fr/;

/**
 * @summary Method used to check if given casUser match student pattern
 * @param {string} casUser cas username
 * @returns {boolean} check result
 */
export function isStudent(casUser: string): boolean {
    if (regex.test(casUser)) {
        return true;
    }
    return false;
}

/**
 * @summary Method used to get email of given cas user
 * @param {string} casUser cas username
 * @returns {string} email
 */
export function getEmail(casUser: string): string {
    return `${casUser}@enib.fr`;
}

/** @summary Macro to set first letter upper case */
function ucFirst(str: string) {
    return str ? `${str[0].toUpperCase()}${str.slice(1)}` : '';
}

/**
 * @summary Method used to guess a mentor name
 * @param {string} casUser cas username
 * @returns {{ firstName: string; lastName: string }} First and last name guessed
 */
export function guessMentorFullName(casUser: string): { firstName: string; lastName: string } {
    const tmp = casUser.split('.');
    if (tmp.length !== 2) {
        throw new Error('Mentor name should use following pattern `firstName.lastName`');
    }
    return {
        firstName: ucFirst(tmp[0]),
        lastName: tmp[1].toUpperCase(),
    };
}

/**
 * @summary Method used to guess a student name
 * @param {string} casUser cas username
 * @returns {{ firstName: string; lastName: string }} First and last name guessed
 */
export function guessStudentFullName(casUser: string): { firstName: string; lastName: string } {
    if (!regex.test(casUser)) {
        throw new Error('Student name should use following pattern `/[a-z][0-9][a-z]{1, 6}/`');
    }
    return {
        firstName: casUser[0].toUpperCase(),
        lastName: casUser.slice(2).toUpperCase(),
    };
}
