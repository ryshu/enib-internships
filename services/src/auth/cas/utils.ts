/** @property {string} regex pattern to identify students */
const regex = /^[a-z][0-9][a-z]{1,6}/;

/**
 * @summary Method used to check if given casUser match student pattern
 * @param {string} casUser cas username
 * @returns {boolean} check result
 */
export function isStudent(casUser: string): boolean {
    if (regex.test(casUser)) {
        return true;
    }
    return false;
}

/**
 * @summary Method used to get email of given cas user
 * @param {string} casUser cas username
 * @returns {string} email
 */
export function getEmail(casUser: string): string {
    return `${casUser}@enib.fr`;
}
