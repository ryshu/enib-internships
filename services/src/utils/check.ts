import {
    ICampaignEntity,
    IFileEntity,
    IBusinessEntity,
    IInternshipEntity,
    IInternshipTypeEntity,
    IMentoringPropositionEntity,
    IMentorEntity,
    IStudentEntity,
} from '../declarations';

export function isInt(t: number): t is number {
    return !!t && !Number.isNaN(Number(t));
}

export function checkPartialBusiness(check: Partial<IBusinessEntity>): check is IBusinessEntity {
    return (
        !!check &&
        !!check.name &&
        !!check.postalCode &&
        !!check.city &&
        !!check.country &&
        !!check.address
    );
}

export function checkPartialCampaign(check: Partial<ICampaignEntity>): check is ICampaignEntity {
    return (
        !!check && !!check.name && !!check.description && !!check.semester && !!check.maxProposition
    );
}

export function checkPartialFile(check: Partial<IFileEntity>): check is IFileEntity {
    return !!check && !!check.name && !!check.path && !!check.type;
}

export function checkPartialInternship(
    check: Partial<IInternshipEntity>,
): check is IInternshipEntity {
    return (
        !!check &&
        !!check.subject &&
        !!check.description &&
        !!check.country &&
        !!check.city &&
        !!check.postalCode &&
        !!check.address &&
        !!check.state
    );
}

export function checkPartialInternshipType(
    check: Partial<IInternshipTypeEntity>,
): check is IInternshipTypeEntity {
    return !!check && !!check.label;
}

export function checkPartialProposition(
    check: Partial<IMentoringPropositionEntity>,
): check is IMentoringPropositionEntity {
    return !!check && !!check.comment;
}

export function checkPartialMentor(check: Partial<IMentorEntity>): check is IMentorEntity {
    return !!check && !!check.firstName && !!check.lastName && !!check.email;
}

export function checkPartialStudent(check: Partial<IStudentEntity>): check is IStudentEntity {
    return !!check && !!check.firstName && !!check.lastName && !!check.email && !!check.semester;
}
