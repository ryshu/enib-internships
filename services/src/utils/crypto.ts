import crypto from 'crypto';

/**
 * @summary Method to encrypt text
 * @param {string} text Text that you want to encrypt
 * @returns {string} encrypted text
 */
export function HelperEncrypt(text: string, key?: string): string {
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv(
        'aes-256-cbc',
        Buffer.from(key ? key : process.env.HASH_KEY),
        iv,
    );
    let encrypted = cipher.update(text);
    encrypted = Buffer.concat([encrypted, cipher.final()]);
    return `${iv.toString('hex')}:${encrypted.toString('hex')}`;
}

/**
 * @summary Method to decrypt text
 * @param {string} text Text that you want to decrpt
 * @returns {string} decrypted text
 */
export function HelperDecrypt(text: string, key?: string): string {
    const textSplited = text.split(':');
    const iv = Buffer.from(textSplited[0], 'hex');
    const encryptedText = Buffer.from(textSplited[1], 'hex');
    const decipher = crypto.createDecipheriv(
        'aes-256-cbc',
        Buffer.from(key ? key : process.env.HASH_KEY),
        iv,
    );
    let decrypted = decipher.update(encryptedText);
    decrypted = Buffer.concat([decrypted, decipher.final()]);
    return decrypted.toString();
}

/**
 * @summary Method use to generate a new random hash key to encode exported file
 * @returns {string} Generated hash key
 */
export function HashKeyGen(): string {
    const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890';
    let key = '';

    for (let i = 0; i < 32; i++) {
        key += chars[Math.floor(Math.random() * 62)];
    }

    return key;
}
