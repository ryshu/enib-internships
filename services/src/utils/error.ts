// TODO: Improve documentations

/**
 * Class used to redefined error instance with encoded error
 */
export class APIError extends Error {
    public code: number;
    public status: number;
    public errors: any;

    constructor(err: string, status: number, code: number) {
        super(err);
        this.name = err;
        this.status = status;
        this.code = code;
        this.errors = null;
    }

    public setErrors(errors: any) {
        this.errors = errors;
    }

    public toJSON() {
        return {
            code: this.code,
            status: this.status,
            errors: this.errors,
            name: this.name,
        };
    }
}
