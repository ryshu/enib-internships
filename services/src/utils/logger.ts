import { createLogger, format, transports } from 'winston';

const winstonLogger = createLogger({
    level: 'info',
    format: format.combine(
        format.timestamp({
            format: 'YYYY-MM-DD HH:mm:ss',
        }),
        format.errors({ stack: true }),
        format.splat(),
        format.json(),
    ),
    transports: [
        new transports.File({ filename: `monitoring.error.log`, level: 'error' }),
        new transports.File({ filename: `monitoring.log` }),
    ],
});

if (process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test') {
    // with the colorized simple format.
    winstonLogger.add(
        new transports.Console({
            format: format.combine(
                format.colorize(),
                format.timestamp({
                    format: 'HH:mm:ss',
                }),
                format.printf((info) => `${info.timestamp} - ${info.level}: ${info.message}`),
            ),
        }),
    );
}

export default winstonLogger;
