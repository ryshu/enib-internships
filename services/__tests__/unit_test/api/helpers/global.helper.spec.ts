import httpStatus from 'http-status-codes';

import { APIError } from '../../../../src/utils/error';

import {
    checkContent,
    checkArrayContent,
    UNPROCESSABLE_ENTITY,
    BAD_REQUEST_VALIDATOR,
} from '../../../../src/api/helpers/global.helper';

describe('checkContent', () => {
    it('checkContent_content_true', () => {
        const mockNext = jest.fn();

        const RESULT = checkContent('content', mockNext);
        expect(RESULT).toBe(true);
        expect(mockNext.mock.calls.length).toBe(0);
    });

    it('checkContent_noContent_false+next', () => {
        const mockNext = jest.fn();

        const RESULT = checkContent(null, mockNext);
        expect(RESULT).toBe(false);
        expect(mockNext.mock.calls.length).toBe(1);
        expect(mockNext.mock.calls[0][0]).toEqual(
            new APIError(
                httpStatus.getStatusText(httpStatus.NO_CONTENT),
                httpStatus.NO_CONTENT,
                11104,
            ),
        );
    });
});

describe('checkArrayContent', () => {
    it('checkArrayContent_content_true', () => {
        const mockNext = jest.fn();

        const RESULT = checkArrayContent(['content'], mockNext);
        expect(RESULT).toBe(true);
        expect(mockNext.mock.calls.length).toBe(0);
    });

    it('checkArrayContent_noContent_false+next', () => {
        const mockNext = jest.fn();

        const RESULT = checkArrayContent([], mockNext);
        expect(RESULT).toBe(false);
        expect(mockNext.mock.calls.length).toBe(1);
        expect(mockNext.mock.calls[0][0]).toEqual(
            new APIError(
                httpStatus.getStatusText(httpStatus.NO_CONTENT),
                httpStatus.NO_CONTENT,
                11104,
            ),
        );
    });
});

describe('UNPROCESSABLE_ENTITY', () => {
    it('UNPROCESSABLE_ENTITY_APIError', () => {
        const mockNext = jest.fn();
        const APIerr = new APIError(
            httpStatus.getStatusText(httpStatus.NO_CONTENT),
            httpStatus.NO_CONTENT,
            11104,
        );

        UNPROCESSABLE_ENTITY(mockNext, APIerr);
        expect(mockNext.mock.calls.length).toBe(1);
        expect(mockNext.mock.calls[0][0]).toEqual(APIerr);
    });

    it('UNPROCESSABLE_ENTITY_Other', () => {
        const mockNext = jest.fn();
        const APIerr = new APIError(
            httpStatus.getStatusText(httpStatus.UNPROCESSABLE_ENTITY),
            httpStatus.UNPROCESSABLE_ENTITY,
            10101,
        );

        UNPROCESSABLE_ENTITY(mockNext, 'APIerr');
        expect(mockNext.mock.calls.length).toBe(1);
        expect(mockNext.mock.calls[0][0]).toEqual(APIerr);
    });
});

describe('BAD_REQUEST_VALIDATOR', () => {
    it('BAD_REQUEST_VALIDATOR_MockedExpressValidatorError', () => {
        const mockNext = jest.fn();
        const err = {
            array() {
                return 'ERROR';
            },
        };

        BAD_REQUEST_VALIDATOR(mockNext, err);
        expect(mockNext.mock.calls.length).toBe(1);
        expect(mockNext.mock.calls[0][0]).toMatchSnapshot();
    });
});
