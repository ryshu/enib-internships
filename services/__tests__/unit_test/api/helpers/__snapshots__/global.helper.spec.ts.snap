// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BAD_REQUEST_VALIDATOR BAD_REQUEST_VALIDATOR_MockedExpressValidatorError 1`] = `[BAD REQUEST: BAD REQUEST]`;
